lib/libmbedtls/core/des3_cmac.c

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2018, ARM Limited
+ * Copyright (C) 2019, Linaro Limited
+ * Copyright (C) 2021, SumUp Services GmbH
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include <kernel/panic.h>
+#include <mbedtls/cipher.h>
+#include <mbedtls/cmac.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <utee_defines.h>
+#include <util.h>
+
+struct mbed_des3_cmac_ctx {
+	struct crypto_mac_ctx mac_ctx;
+	mbedtls_cipher_context_t cipher_ctx;
+};
+
+static const struct crypto_mac_ops mbed_des3_cmac_ops;
+
+static struct mbed_des3_cmac_ctx *to_des3_cmac_ctx(struct crypto_mac_ctx *ctx)
+{
+	assert(ctx && ctx->ops == &mbed_des3_cmac_ops);
+
+	return container_of(ctx, struct mbed_des3_cmac_ctx, mac_ctx);
+}
+
+static TEE_Result mbed_des3_cmac_init(struct crypto_mac_ctx *ctx,
+				 const uint8_t *key, size_t len)
+{
+	struct mbed_des3_cmac_ctx *c = to_des3_cmac_ctx(ctx);
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+
+	cipher_info = mbedtls_cipher_info_from_values(MBEDTLS_CIPHER_ID_3DES,
+						      len * 8,
+						      MBEDTLS_MODE_ECB);
+	if (!cipher_info)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (mbedtls_cipher_setup_info(&c->cipher_ctx, cipher_info))
+		return TEE_ERROR_BAD_STATE;
+
+	if (mbedtls_cipher_cmac_reset(&c->cipher_ctx))
+		return TEE_ERROR_BAD_STATE;
+
+	if (mbedtls_cipher_cmac_starts(&c->cipher_ctx, key, len * 8))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_des3_cmac_update(struct crypto_mac_ctx *ctx,
+				   const uint8_t *data, size_t len)
+{
+	if (mbedtls_cipher_cmac_update(&to_des3_cmac_ctx(ctx)->cipher_ctx,
+				       data, len))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_des3_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+				  size_t len)
+{
+	struct mbed_des3_cmac_ctx *c = to_des3_cmac_ctx(ctx);
+	uint8_t block_digest[TEE_DES_BLOCK_SIZE] = { 0 };
+	uint8_t *tmp_digest = NULL;
+
+	if (len == 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (len < sizeof(block_digest))
+		tmp_digest = block_digest; /* use a tempory buffer */
+	else
+		tmp_digest = digest;
+
+	if (mbedtls_cipher_cmac_finish(&c->cipher_ctx, tmp_digest))
+		return TEE_ERROR_BAD_STATE;
+
+	if (len < sizeof(block_digest))
+		memcpy(digest, tmp_digest, len);
+
+	return TEE_SUCCESS;
+}
+
+static void mbed_des3_cmac_free_ctx(struct crypto_mac_ctx *ctx)
+{
+	struct mbed_des3_cmac_ctx *c = to_des3_cmac_ctx(ctx);
+
+	mbedtls_cipher_free(&c->cipher_ctx);
+	free(c);
+}
+
+static void mbed_des3_cmac_copy_state(struct crypto_mac_ctx *dst_ctx,
+				 struct crypto_mac_ctx *src_ctx)
+{
+	struct mbed_des3_cmac_ctx *src = to_des3_cmac_ctx(src_ctx);
+	struct mbed_des3_cmac_ctx *dst = to_des3_cmac_ctx(dst_ctx);
+
+	if (mbedtls_cipher_clone(&dst->cipher_ctx, &src->cipher_ctx))
+		panic();
```

Comment:
Hmmm that's a bit violent! :fearful:

I think we can do better than crash the whole TEE... would it be possible to set the destination context to a known invalid state so that an error would be reported a bit later to the caller? Not very clean but better than `panic()`.

Another option would be to change the prototype of the function and return a `TEE_Result`. Unless it involves changing a lot of code that may be the best choice.

Reply:
I just followed the pattern from the file which is already part of OP-TEE.
Please take a look at https://github.com/OP-TEE/optee_os/blob/master/lib/libmbedtls/core/aes_cmac.c#L106-L108

Reply:
So if we want to change it here we should also fix AES CMAC code as well.

Reply:
Oh, I see... perhaps it is a "can't happen" thing because `mbedtls_cipher_clone()` doesn't allocate anything and is not supposed to fail assuming its input is valid? [/me checks the MBed TLS code...] Yeah that is indeed the case, so this is correct.

Reply:
if can't happen, prefer `assert()` than `panic()`.

Reply:
> if can't happen, prefer `assert()` than `panic()`.

I agree, but then we should update AES-CMAC too. Perhaps better to take in a follow up PR?

Reply:
Sure.
Do you want me to create it?
Sorry, if you were talking to @etienne-lms :)